goals/mcp-kit/history/2026-07-01-p0-verification.md

```
@@ -0,0 +1,212 @@
+areas this packet depends on. Per the task instruction ("verify against what
+the workspace actually resolves"), **`node_modules/effect/src/unstable/ai/*`
+is the authority used below**, not the `.repos/effect-v4` subtree. This
+tree-divergence itself is worth a maintenance note (see Open Risks) but does
+not affect any claim.
+
+## Verified Internals
+
+| Claim | Current anchor (resolved `node_modules/effect`) | Verdict | Decisive quote |
+| --- | --- | --- | --- |
+| (a) `failureMode:"return"` typed failures ship as `CallToolResult` with `isError:false` | `McpServer.ts:717-728` — **unchanged, exact line match** | CONFIRMED | `Effect.matchCause({ onFailure: (cause) => new CallToolResult({ isError: true, ... }), onSuccess: (result: any) => new CallToolResult({ isError: false, structuredContent: ..., content: [{ type: "text", text: JSON.stringify(result.encodedResult) }] }) })` (McpServer.ts:717-734). Mechanism confirmed in `Toolkit.ts:240-242`: `tool.failureMode === "return"` folds the failure schema into the *success* result union (`Schema.Union([tool.successSchema, tool.failureSchema, AiError.AiError])`), so a typed "return" failure completes the handler's Effect as a success and lands in the `onSuccess` branch → `isError:false`, JSON in `content[].text`. |
+| (b) `McpSchema.EnabledWhen` filters `tools/list` only; `tools/call` dispatch does not re-check it | `callTool` (the `tools/call` dispatcher): `McpServer.ts:255-262` — **exact line match**. `tools/list` handler entry: `McpServer.ts:1451` (`"tools/list": (_, { client, headers }) =>`) — **off by one line from the prior `:1450` citation** (negligible; caused by unrelated line-count drift elsewhere in the 1524-line file, not a semantic change). The actual `EnabledWhen` check lives in the shared `filterByClient` helper at `McpServer.ts:1490-1512`, check itself at `:1506-1507`. | CONFIRMED (line-1 drift, non-invalidating) | `callTool: (request) => Effect.suspend(...): { const handle = toolMap.get(request.name); if (!handle) { return Effect.fail(...) } return handle(request.arguments) }` (McpServer.ts:255-262) — no `EnabledWhen`/`Context` lookup anywhere in this path. Contrast `tools/list`: `"tools/list": (_, { client, headers }) => ... tools: filterByClient(initialized, server.tools, "tool")` (McpServer.ts:1451/1455), and `filterByClient`: `const enabledWhen = Context.getOrUndefined(item.annotations, EnabledWhen); if (!enabledWhen || enabledWhen(client)) { out.push(item[prop]) }` (McpServer.ts:1506-1507). |
```

markdownlint-cli2 (0.22.1)

[warning] Table column count
Expected: 4; Actual: 6; Too many cells, extra data will be missing

(MD056, table-column-count)